extra/lrshift~/lrshift~.c

```diff
@@ -7,68 +7,67 @@ static t_class *lrshift_tilde_class;
 typedef struct _lrshift_tilde
 {
     t_object x_obj;
-    int x_n;
-    t_float x_f;
+    t_float  x_n;
+    t_float  x_f;
 } t_lrshift_tilde;
 
-static t_int *leftshift_perform(t_int *w)
+static t_int *lrshift_perform(t_int *w)
 {
-    t_sample *in = (t_sample *)(w[1]);
-    t_sample *out= (t_sample *)(w[2]);
-    int n = (int)(w[3]);
-    int shift = (int)(w[4]);
-    in += shift;
-    n -= shift;
-    while (n--)
-        *out++ = *in++;
-    while (shift--)
-        *out++ = 0;
-    return (w+5);
-}
+    t_lrshift_tilde *x = (t_lrshift_tilde *) w[1];
+    t_sample *in = (t_sample *)(w[2]);
+    t_sample *out= (t_sample *)(w[3]);
+    int n = (int)(w[4]);
+    int shift = (int)x->x_n;
+    t_sample *ptr,
+        *out_end = out + n - 1,
+        *in_end = in + n - 1;
 
-static t_int *rightshift_perform(t_int *w)
-{
-    t_sample *in = (t_sample *)(w[1]);
-    t_sample *out= (t_sample *)(w[2]);
-    int n = (int)(w[3]);
-    int shift = (int)(w[4]);
-    n -= shift;
-    in -= shift;
-    while (n--)
-        *--out = *--in;
-    while (shift--)
-        *--out = 0;
-    return (w+5);
+    /* The loops are written this way to avoid "read after write" errors
+    when *in and *out point to the same array */
+    if (shift >= 0)  /* shift left, shift is a postive int or zero */
+    {
+        if (shift > n)
+            shift = n;
+        for (ptr = in + shift; ptr <= in_end;)
+            *out++ = *ptr++;
+        while (shift--)
+            *out++ = 0.0f;
+    }
+    else /* shift right, shift is a negative int */
+    {
+        if (shift < -n)
+            shift = -n;
+        t_sample *out_end = out + n-1;
+        for (ptr = in_end + shift; ptr >= in;)
+            *out_end-- = *ptr--;
+        while (shift++)
+            *out_end-- = 0.0f;
+    }
+
+    return(w+5);
 }
 
 static void lrshift_tilde_dsp(t_lrshift_tilde *x, t_signal **sp)
 {
-    int n = sp[0]->s_n;
-    int shift = x->x_n;
-    if (shift > n)
-        shift = n;
-    if (shift < -n)
-        shift = -n;
-    if (shift < 0)
-        dsp_add(rightshift_perform, 4,
-            sp[0]->s_vec + n, sp[1]->s_vec + n, (t_int)n, (t_int)(-shift));
-    else dsp_add(leftshift_perform, 4,
-            sp[0]->s_vec, sp[1]->s_vec, (t_int)n, (t_int)shift);
+    t_int n = sp[0]->s_n;
+    dsp_add(lrshift_perform, 4, x,
+        sp[0]->s_vec, sp[1]->s_vec, n);
```

Comment:
By convention, we would rather write
```
    dsp_add(lrshift_perform, 4, x,
        sp[0]->s_vec, sp[1]->s_vec, (t_int)sp[0]->s_n);
```
The advantage is that it makes the (very important) cast to `t_int` more visible.